### DISCUSSION TITLE
Allow server list sorting

### DISCUSSION THREAD

**Discussion by sxiii** [jamulussoftware/jamulus#1146] [Ideas]
Asc/desc sort function into server list by number of users / location / ping / server name

I would like to choose server based on different criteria.

**Comment by ann0see** [comment]
Maybe this is already possible with the filter option next to the list selection?

**Comment by dcorson-ticino-com** [comment]
For the moment the server list is only sorted by ping time.
The filter function reduces the list to servers where the filter string is found in one of the fields.
Personally I have never felt the need to sort by anything other than by ping, the filter works well for special occasions, but it would certainly be possible if we decide that is desirable.

**Comment by gilgongo** [comment]
Congratulations @sxiii ! You are the first person to use our new Issue reporting template. 

Since this is I think a pretty clear request (and sorting is in fact implemented when --showallservers is on I think), I propose we keep it here. If anyone wants to pull it into a protracted discussion (eg if this feature would break something) they can do, at which point we can move it out to Discussions if needed.

**Comment by dcorson-ticino-com** [comment]
I correct myself.  
Because of a apparent code error, the code doesn't do what the commentary says, you can sort on all but the server name when you use the  --showallservers command line option.
The disadvantage is that you see all the servers the central server knows about, even those that do not respond to the pings of your computer.
A registration number is also prepended to the server name which is why the sort on server name does not work.

**Comment by gilgongo** [comment]
OK so this isn't as straightforward as making sorting available in the non-showallservers condition then. And there may be other thing. Moving to a discussion.

**Comment by dingodoppelt** [comment]
> Asc/desc sort function into server list by number of users / location / ping / server name
> I would like to choose server based on different criteria.

https://github.com/dingodoppelt/jamulus/tree/sortableserverlist

Hi, I fixed it in my fork. Give it a try :)

**Reply by gilgongo** [reply]
Doesn't seem to want to sort columns for me on mouse click, if that's what you mean? 

![image](https://user-images.githubusercontent.com/4561747/109494011-38603400-7a85-11eb-8ae0-6df6fb78c99b.png)

**Reply by dingodoppelt** [reply]
have you checked out the correct branch? "git branch" should tell you. otherwise do "git clone -b sortableserverlist https://github.com/dingodoppelt/jamulus"

**Reply by gilgongo** [reply]
Ah OK that's more like it. Seems to work for me (Ubuntu 20.04). What order does it start in though? It should really default to ping time but doesn't with me at least.

**Reply by dcorson-ticino-com** [reply]
I haven't looked at the code, but remember that when sorted according to the ping time it is sorted according to the minimum ping time seen for a channel (to reduce jumping around of the list), which is in a hidden column, and must be sorted periodically.  For the other categories I would think that periodic sorting is not necessary, but just after a click in the heading is enough.

**Reply by gilgongo** [reply]
Having it default to sorted by ping times is best though I think, because most of the time that's what you're interested in for jamming at least. BTW is the "pause on mouse over" disabled in this build? Seem so.

**Reply by dingodoppelt** [reply]
>  BTW is the "pause on mouse over" disabled in this build? Seem so.

yes, because it really bugged me the most. i changed it to sort by musicians per default because i found myself clicking the "musicians" for sorting anyway. and it works around the list jumping. if you want to join a server do it the old fashioned way: click a server once, then hit "enter" to join that server. no need for doubleclick :)

**Reply by gilgongo** [reply]
Ah OK, and yes that mouseover feature was a bit odd. I didn't spot that it's sorted by musicians :-) A default sort of ping time sub-sorted by musicians would be pretty good if that was possible.

**Reply by dcorson-ticino-com** [reply]
I don't think this is a good idea for general release.  The ping time is too important to not have it as the major sorting factor.
It would be good if we could have an optional two column sort, like musicians and ping.  Unfortunately I do not see this capability in Qt at the moment, we would have to develop it.
I understand why the mouseover was introduced, but it is something of a hack in use.  We do need to retain at least the use of the minimum ping time to reduce the list jumping.  This is not in the --showallservers code.
In general, this will be much more work than just making the --showallservers case the default.

**Reply by gilgongo** [reply]
@dcorson-ticino-com Nobody's suggesting --showallservers should be the default, are they? We're just talking about making columns sort as is the case in --showallservers.
